### PR TITLE
Deprecate status field in SCA

### PR DESCRIPTION
**Pull request by chemamartinez**
|Related issue|
|---|
|https://github.com/wazuh/wazuh/issues/15695|

## Description

This pull request is the aggregation of https://github.com/wazuh/wazuh/pull/15745 (core) and https://github.com/wazuh/wazuh/pull/15747 (API and framework) to deprecate the `status` field in SCA.

# Core changes

## Description

This pull request deprecates the `status` field in SCA, which was being used to report when a check was not able to be executed. Now, this `not applicable` status is reflected in the `result` field, which can also take the values `passed` and `failed`.

This `not applicable` value is still followed by the `reason` field where it is detailed the cause of the failure when evaluating the check.

In addition, this change only applies to the manager side for now (decoder, alerts, and database). The SCA scanner, the events sent by agents, and the synchronization mechanism between databases still support the `status` field to avoid breaking compatibility with older agents.

The following changes have been applied:
- Deprecate the `status` field in the SCA database: f5384e438b6a1a9ad9b84d18903a3cf7b7acc704
- Deprecate the `status` field in SCA alerts: 37a4b5ce8ff1b54bc0a56ea55b107e4247c2748c
- Migrate the SCA database to the new schema at manager upgrades: f36b8aca027e9fc0296012d77818be23a8b8475f

## Use case

To test the changes and show how it works, the following use case is done:

1. Install Wazuh v4.3.11 to start from a version without changes.
2. Configure an SCA policy that contains a check for every possible result.
<details><summary>sca_results_test_suite.yml</summary>

```yml
policy:
  id: sca_results_test_suite
  file: sca_results_test_suite.yml
  name: Test suite for all the possible results
  description: This document includes rules for testing purposes
################ Expected results
# Passed = 1
# Failed = 1
# Invalid = 1
# Total checks = 3
checks:
 - id: 200100
   title: PASS -- existing file exists
   condition: any
   rules:
     - f:/var/ossec/etc/ossec.conf

 - id: 200101
   title: FAIL -- a non-existing file file should exist
   condition: any
   rules:
     - f:/var/ossec/etc/FAKE.conf

 - id: 200102
   title: INVALID -- File does not exist
   condition: any
   rules:
     - f:/var/ossec/etc/FAKE.conf -> r:^FAKE
```

SCA configuration:
```xml
  <sca>
    <enabled>yes</enabled>
    <scan_on_start>yes</scan_on_start>
    <interval>12h</interval>
    <skip_nfs>yes</skip_nfs>
    <policies>
      <policy>/root/sca_results_test_suite.yml</policy>
    </policies>
  </sca>
```

</details>

3. Here we can see the alerts before the changes:

<details><summary>SCA alerts in v4.3.11</summary>

```console
** Alert 1671725396.0: - sca,gdpr_IV_35.7.d,pci_dss_2.2,nist_800_53_CM.1,tsc_CC7.1,tsc_CC7.2,
2022 Dec 22 17:09:56 wazuh->sca
Rule: 19008 (level 3) -> 'Test suite for all the possible results: PASS -- existing file exists'
{"type":"check","id":185613349,"policy":"Test suite for all the possible results","policy_id":"sca_results_test_suite","check":{"id":200100,"title":"PASS -- existing file exists","rules":["f:/var/ossec/etc/ossec.conf"],"condition":"any","file":"/var/ossec/etc/ossec.conf","result":"passed"}}
sca.type: check
sca.scan_id: 185613349
sca.policy: Test suite for all the possible results
sca.check.id: 200100
sca.check.title: PASS -- existing file exists
sca.check.file: ["/var/ossec/etc/ossec.conf"]
sca.check.result: passed

** Alert 1671725396.745: - sca,gdpr_IV_35.7.d,pci_dss_2.2,nist_800_53_CM.1,tsc_CC7.1,tsc_CC7.2,
2022 Dec 22 17:09:56 wazuh->sca
Rule: 19007 (level 7) -> 'Test suite for all the possible results: FAIL -- a non-existing file file should exist'
{"type":"check","id":185613349,"policy":"Test suite for all the possible results","policy_id":"sca_results_test_suite","check":{"id":200101,"title":"FAIL -- a non-existing file file should exist","rules":["f:/var/ossec/etc/FAKE.conf"],"condition":"any","file":"/var/ossec/etc/FAKE.conf","result":"failed"}}
sca.type: check
sca.scan_id: 185613349
sca.policy: Test suite for all the possible results
sca.check.id: 200101
sca.check.title: FAIL -- a non-existing file file should exist
sca.check.file: ["/var/ossec/etc/FAKE.conf"]
sca.check.result: failed

** Alert 1671725396.1540: - sca,gdpr_IV_35.7.d,pci_dss_2.2,nist_800_53_CM.1,tsc_CC7.1,tsc_CC7.2,
2022 Dec 22 17:09:56 wazuh->sca
Rule: 19009 (level 3) -> 'Test suite for all the possible results: INVALID -- File does not exist'
{"type":"check","id":185613349,"policy":"Test suite for all the possible results","policy_id":"sca_results_test_suite","check":{"id":200102,"title":"INVALID -- File does not exist","rules":["f:/var/ossec/etc/FAKE.conf -> r:^FAKE"],"condition":"any","file":"/var/ossec/etc/FAKE.conf","status":"Not applicable","reason":"Could not open file '/var/ossec/etc/FAKE.conf'"}}
sca.type: check
sca.scan_id: 185613349
sca.policy: Test suite for all the possible results
sca.check.id: 200102
sca.check.title: INVALID -- File does not exist
sca.check.file: ["/var/ossec/etc/FAKE.conf"]
sca.check.status: Not applicable
sca.check.reason: Could not open file '/var/ossec/etc/FAKE.conf'

** Alert 1671725403.2683: - sca,gdpr_IV_35.7.d,pci_dss_2.2,nist_800_53_CM.1,tsc_CC7.1,tsc_CC7.2,
2022 Dec 22 17:10:03 wazuh->sca
Rule: 19003 (level 5) -> 'SCA summary: Test suite for all the possible results: Score less than 80% (50)'
{"type":"summary","scan_id":185613349,"name":"Test suite for all the possible results","policy_id":"sca_results_test_suite","file":"sca_results_test_suite.yml","description":"This document includes rules for testing purposes","passed":1,"failed":1,"invalid":1,"total_checks":3,"score":50,"start_time":1671725391,"end_time":1671725391,"hash":"5e68cc97df4c3592165b74578364dc6421f899797bd99ab5e2a7eaadf1117267","hash_file":"a304d4f04495becaa401779d608a49dd2974eddb2d2ffb6b273ad16a370be70b","force_alert":"1"}
sca.type: summary
sca.scan_id: 185613349
sca.policy: Test suite for all the possible results
sca.description: This document includes rules for testing purposes
sca.policy_id: sca_results_test_suite
sca.passed: 1
sca.failed: 1
sca.invalid: 1
sca.total_checks: 3
sca.score: 50
sca.file: sca_results_test_suite.yml
```
</details>

4. Here we have the content of the database:
```
sqlite> select id,result,status,reason from sca_check;
200100|passed||
200101|failed||
200102||Not applicable|Could not open file '/var/ossec/etc/FAKE.conf'
```
We can see that the last check stores the `Not applicable` result in the `status`field.
```
sqlite> select policy_id,hash from sca_scan_info;
sca_results_test_suite|5e68cc97df4c3592165b74578364dc6421f899797bd99ab5e2a7eaadf1117267
```
This is the checksum of the results for that policy. It must be the same once the changes are applied.

5. In this point, **we upgrade the manager to v4.4.0**. Now, the database looks like:

```
sqlite> select * from metadata;
db_version|10
sqlite> select id,result,reason from sca_check;
200100|passed|
200101|failed|
200102|not applicable|Could not open file '/var/ossec/etc/FAKE.conf'
sqlite> select status from sca_check;
Error: no such column: status
```
As we can see, the database has been updated. In addition, the `status`field doesn't exist in the `sca_check`table. Now all the checks store their statuses in the `result`field.

```
sqlite> select policy_id,hash from sca_scan_info;
sca_results_test_suite|5e68cc97df4c3592165b74578364dc6421f899797bd99ab5e2a7eaadf1117267
```
We also verify that the integrity hash is the same as before the upgrade.

6. Finally, we will test the new alerts by forcing a change in the policy results. To do this we create the following file:
```
# touch /var/ossec/etc/FAKE.conf
```

Now, the check 200101 switches from `failed`to `passed`because the file exists. In the same way, the check 200102 goes from `not applicable` to `failed` because the file exists but the content doesn't match the rule.

Here we can see the generated alerts in the next scan.

<details><summary>SCA alerts in v4.4.0</summary>

```console
** Alert 1671726070.5295: - sca,gdpr_IV_35.7.d,pci_dss_2.2,nist_800_53_CM.1,tsc_CC7.1,tsc_CC7.2,
2022 Dec 22 17:21:10 wazuh->sca
Rule: 19010 (level 3) -> 'Test suite for all the possible results: FAIL -- a non-existing file file should exist: Status changed from failed to passed'
{"type":"check","id":441062567,"policy":"Test suite for all the possible results","policy_id":"sca_results_test_suite","check":{"id":200101,"title":"FAIL -- a non-existing file file should exist","rules":["f:/var/ossec/etc/FAKE.conf"],"condition":"any","file":"/var/ossec/etc/FAKE.conf","result":"passed"}}
sca.type: check
sca.scan_id: 441062567
sca.policy: Test suite for all the possible results
sca.check.id: 200101
sca.check.title: FAIL -- a non-existing file file should exist
sca.check.file: ["/var/ossec/etc/FAKE.conf"]
sca.check.result: passed
sca.check.previous_result: failed

** Alert 1671726070.6163: - sca,gdpr_IV_35.7.d,pci_dss_2.2,nist_800_53_CM.1,tsc_CC7.1,tsc_CC7.2,
2022 Dec 22 17:21:10 wazuh->sca
Rule: 19014 (level 9) -> 'Test suite for all the possible results: INVALID -- File does not exist: Status changed from 'not applicable' to failed'
{"type":"check","id":441062567,"policy":"Test suite for all the possible results","policy_id":"sca_results_test_suite","check":{"id":200102,"title":"INVALID -- File does not exist","rules":["f:/var/ossec/etc/FAKE.conf -> r:^FAKE"],"condition":"any","file":"/var/ossec/etc/FAKE.conf","result":"failed"}}
sca.type: check
sca.scan_id: 441062567
sca.policy: Test suite for all the possible results
sca.check.id: 200102
sca.check.title: INVALID -- File does not exist
sca.check.file: ["/var/ossec/etc/FAKE.conf"]
sca.check.result: failed
sca.check.previous_result: not applicable

** Alert 1671726077.7259: - sca,gdpr_IV_35.7.d,pci_dss_2.2,nist_800_53_CM.1,tsc_CC7.1,tsc_CC7.2,
2022 Dec 22 17:21:17 wazuh->sca
Rule: 19003 (level 5) -> 'SCA summary: Test suite for all the possible results: Score less than 80% (66)'
{"type":"summary","scan_id":441062567,"name":"Test suite for all the possible results","policy_id":"sca_results_test_suite","file":"sca_results_test_suite.yml","description":"This document includes rules for testing purposes","passed":2,"failed":1,"invalid":0,"total_checks":3,"score":66.666671752929688,"start_time":1671726065,"end_time":1671726065,"hash":"2c1982d54d3b8f5bb28cd6227d2e52f9db22abde46ec2a449e7131e5ce847c95","hash_file":"a304d4f04495becaa401779d608a49dd2974eddb2d2ffb6b273ad16a370be70b","force_alert":"1"}
sca.type: summary
sca.scan_id: 441062567
sca.policy: Test suite for all the possible results
sca.description: This document includes rules for testing purposes
sca.policy_id: sca_results_test_suite
sca.passed: 2
sca.failed: 1
sca.invalid: 0
sca.total_checks: 3
sca.score: 66
sca.file: sca_results_test_suite.yml
```
</details>

# API and framework changes

## Description

This PR adapts the framework and API to the changes in the `sca_check` database. From now on, the response field `status` will not appear nor be eligible as a query filter.

### Standard query (`result=not applicable` & `limit=1`)

```json
{
  "data": {
    "affected_items": [
      {
        "description": "The cramfs filesystem type is a compressed read-only Linux filesystem embedded in small footprint systems. A cramfs image can be used without having to first decompress the image.",
        "rationale": "Removing support for unneeded filesystem types reduces the local attack surface of the server. If this filesystem type is not needed, disable it.",
        "condition": "all",
        "reason": "Invalid path or wrong permissions to run command 'modprobe -n -v cramfs'",
        "policy_id": "cis_ubuntu20-04",
        "title": "Ensure mounting of cramfs filesystems is disabled.",
        "remediation": "Edit or create a file in the /etc/modprobe.d/ directory ending in .conf Example: vim /etc/modprobe.d/cramfs.conf and add the following line: install cramfs /bin/true Run the following command to unload the cramfs module: # rmmod cramfs",
        "result": "not applicable",
        "id": 19000,
        "command": "modprobe -n -v cramfs,lsmod",
        "compliance": [
          {
            "key": "cis",
            "value": "1.1.1.1"
          },
          {
            "key": "cis_csc_v7",
            "value": "5.1"
          },
          {
            "key": "cmmc_v2.0",
            "value": "AC.1.002,CM.2.061,SC.3.180"
          },
          {
            "key": "iso_27001-2013",
            "value": "A.8.1.3,A.14.2.5"
          },
          {
            "key": "mitre_techniques",
            "value": "T1110,T1003,T1081,T1097,T1178,T1072,T1067,T1495,T1019,T1177,T1485,T1486,T1491,T1488,T1487,T1490,T1146,T1148,T1015,T1133,T1200,T1076,T1051,T1176,T1501,T1087,T1098,T1139,T1197,T1092,T1136,T1011,T1147,T1130,T1174,T1053,T1166,T1206,T1503,T1214,T1187,T1208,T1142,T1075,T1201,T1145,T1184,T1537,T1078,T1077,T1134,T1017,T1088,T1175,T1190,T1210,T1525,T1215,T1086,T1055,T1505,T1035,T1218,T1169,T1100,T1047,T1084,T1028,T1156,T1196,T1530,T1089,T1073,T1157,T1054,T1070,T1037,T1036,T1096,T1034,T1150,T1504,T1494,T1489,T1198,T1165,T1492,T1080,T1209,T1112,T1058,T1173,T1137,T1539,T1535,T1506,T1138,T1044,T1199"
          },
          {
            "key": "nist_sp_800-53",
            "value": "AU-2,CM-1,CM-2,CM-6,CM-7,IA-5,IA-6,SC-20,SC-21"
          },
          {
            "key": "pci_dss_v3.2.1",
            "value": "2.2"
          }
        ],
        "rules": [
          {
            "type": "command",
            "rule": "c:modprobe -n -v cramfs -> r:^install /bin/true|^install /bin/false"
          },
          {
            "type": "numeric",
            "rule": "not c:lsmod -> r:cramfs"
          }
        ]
      }
    ],
    "total_affected_items": 31,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "All selected sca/policy information was returned",
  "error": 0
}
```

### Trying to use the `status` filter

```json
{
  "title": "Bad Request",
  "detail": "Field does not exist.: Available fields: policy_id, id, title, description, rationale, remediation, file, process, directory, registry, command, references, result, reason, condition, compliance.key, compliance.value, rules.type, rules.rule. Field: status",
  "dapi_errors": {
    "master-node": {
      "error": "Field does not exist.: Available fields: policy_id, id, title, description, rationale, remediation, file, process, directory, registry, command, references, result, reason, condition, compliance.key, compliance.value, rules.type, rules.rule. Field: status"
    }
  },
  "error": 1408
}
```

### CODE REVIEW

**Review by vicferpoy — APPROVED** [review]
API and framework related code LGTM.

File: src/wazuh_db/schema_upgrade_v10.sql [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+   condition TEXT
+);
+
+INSERT INTO _sca_check SELECT scan_id, id, policy_id, title, description, rationale, remediation, file, process, directory, registry, command, `references`, CASE WHEN result <> '' THEN result ELSE 'not applicable' END AS result, reason, condition FROM sca_check;
+
+DROP TABLE IF EXISTS sca_check;
+
+ALTER TABLE _sca_check RENAME TO sca_check;
+
+CREATE INDEX IF NOT EXISTS policy_id_index ON sca_check (policy_id);
```
Comment (vikman90) [inline src/wazuh_db/schema_upgrade_v10.sql:36]: I guess we can simplify this operation with something like this:

```suggestion
UPDATE sca_check SET result='not applicable' WHERE status='not applicable';
ALTER TABLE sca_check DROP COLUMN status;
```

This avoids moving all the table data and reindexing it.

**Review by vikman90 — CHANGES_REQUESTED** [review]

File: src/wazuh_db/schema_upgrade_v10.sql (hunk 1 above)
Comment (chemamartinez) [reply src/wazuh_db/schema_upgrade_v10.sql]: Done and tested f72ac75df7288169883ff5012c6c9017597b4ff8. Same results as before:
```
sqlite> select status from sca_check;
Error: no such column: status
sqlite> select result,reason from sca_check;
passed|
failed|
not applicable|Could not open file '/var/ossec/etc/FAKE.conf'
```
Thanks for the suggestions!

**Review by vikman90 — APPROVED** [review]

**Comment by chemamartinez** [comment]
QA team also approved these changes: https://github.com/wazuh/wazuh/pull/15745#issuecomment-1378828961

**Pull request merged**